src/logging/container.ts

```diff
@@ -0,0 +1,98 @@
+import * as es from 'event-stream';
+import { EventEmitter } from 'events';
+import * as _ from 'lodash';
+import * as Stream from 'stream';
+import StrictEventEmitter from 'strict-event-emitter-types';
+
+import Docker from '../lib/docker-utils';
+
+export interface ContainerLog {
+	message: string;
+	timestamp: number;
+	isStdout: boolean;
+}
+
+interface LogsEvents {
+	log: ContainerLog;
+	closed: void;
+	error: Error;
+}
+
+type LogsEventEmitter = StrictEventEmitter<EventEmitter, LogsEvents>;
+
+export class ContainerLogs extends (EventEmitter as {
+	new (): LogsEventEmitter;
+}) {
+	public constructor(public containerId: string, private docker: Docker) {
+		super();
+	}
+
+	public async attach(lastSentTimestamp: number) {
+		const logOpts = {
+			follow: true,
+			timestamps: true,
+			since: Math.floor(lastSentTimestamp / 1000),
+		};
+		const stdoutLogOpts = _.merge({ stdout: true, stderr: false }, logOpts);
+		const stderrLogOpts = _.merge({ stderr: true, stdout: false }, logOpts);
+
+		const container = this.docker.getContainer(this.containerId);
+		const stdoutStream = await container.logs(stdoutLogOpts);
+		const stderrStream = await container.logs(stderrLogOpts);
+
+		[[stdoutStream, true], [stderrStream, false]].forEach(
+			([stream, isStdout]: [Stream.Readable, boolean]) => {
+				stream
+					.on('error', err => {
+						this.emit(
+							'error',
+							new Error(`Error on container logs: ${err} ${err.stack}`),
+						);
+					})
+					.pipe(es.split())
+					.on('data', (logBuf: Buffer | string) => {
+						if (_.isString(logBuf)) {
+							logBuf = Buffer.from(logBuf);
+						}
+						const logMsg = ContainerLogs.extractMessage(logBuf);
+						if (logMsg != null) {
+							this.emit('log', _.merge({ isStdout }, logMsg));
+						}
+					})
+					.on('error', err => {
+						this.emit(
+							'error',
+							new Error(`Error on container logs: ${err} ${err.stack}`),
+						);
+					})
+					.on('end', () => this.emit('closed'));
+			},
+		);
+	}
+
+	private static extractMessage(
+		msgBuf: Buffer,
+	): { message: string; timestamp: number } | null {
+		// Non-tty message format from:
+		// https://docs.docker.com/engine/api/v1.30/#operation/ContainerAttach
+		if (msgBuf[0] in [0, 1, 2] && _.every(msgBuf.slice(1, 7), c => c === 0)) {
```

Comment:
This is typescript, `in` doesn't mean the same thing as coffeescript so this needs to be
```suggestion
		if (_.includes([0, 1, 2], msgBuf[0]) && _.every(msgBuf.slice(1, 7), c => c === 0)) {
```

Reply:
Good catch! This is actually code from a while ago, so that bug has been there for ages!